proxy.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * @internal Symbol used to track if the object has been mutated
+ */
+export const _MUTATED = Symbol("mutated");
+
+/**
+ * @internal Symbol used to track if the object is being tracked
+ */
+export const _TRACKED = Symbol("tracked");
+
+type Mutatable<T> = T & { [_MUTATED]?: boolean; [_TRACKED]?: boolean };
+
+// for Set, Map, and Array
+const MUTATION_METHODS = new Set([
+    "set",
+    "add",
+    "delete",
+    "clear",
+    "push",
+    "pop",
+    "shift",
+    "unshift",
+    "splice",
+    "sort",
+    "reverse",
+    "fill",
+    "copyWithin",
+]);
+
+/**
+ * @internal Creates a proxy wrapper around an object to track mutations
+ * @param object - The target object to proxify
+ * @returns Proxied object that tracks mutations
+ */
+export function trackMutations<T>(
+    object: T,
+): Mutatable<T> {
+    if (typeof object !== "object" || object === null) {
+        return object as Mutatable<T>;
+    }
+
+    const proxyCache = new WeakMap();
```

Comment:
How hard would it be to tie the cache to a property session instead? That would feel like a more obvious strategy to manage the lifetime of these objects in memory

Reply:
I don't think it'd be that hard, but it's not obvious to me why we'd want to do it, sorry.

Reply:
I find it hard to reason about the lifecycle of memory when there are different references to the objects, and only some of them prevent it from being garbage-collected. This is especially so if there is an obvious way to structure that code such that the memory is directly tied to the stack, which makes it very obvious how long the objects live. But if you have strong opinions about keeping it in a `WeakMap`, then I'll be able to wrap my head around it :)